### PR TITLE
Community latest telegram messages

### PR DESCRIPTION
**Pull request by ailisp**
Address #78. This uses the official telegram widget (iframe) to display individual message and I setup an aws lambda service to get the latest 10 message IDs from telegram groups: https://github.com/ailisp/cg-msg-counter/blob/main/index.js. There is no frontend-only solution to retrieve the message IDs.

Preview: https://near.social/#/bo.near/widget/gigs-board.pages.Feed?nearDevGovGigsContractAccountId=devgovgigs.near

Note that, due to #81 is not there, this PR uses hard coded telegram groups.

@ori-near ~I have trouble in layout these telegram messages in a beautiful way (they're iframes from https://t.me/`), may you take a look at the css here?~

There is no pure css way to make every iframe as height as the content. The JS way isn't work because near social VM prohibit direct access to DOM element. Therefore, I will modify the service to return messages as json and render with CSS to try looks close to telegram's UI

### CODE REVIEW

**Comment by ailisp** [comment]
So this problem is surprisingly challenging due to the limitation that combined by telegram API & near social. So far I have experimented the following approaches, but didn't get any of them work on the near social yet:

1. direct iframe of https://t.me message widget. The iframes are not layout well since css can't size the chat message according to its content. JS doesn't work because near social prevent DOM access.
2. host my static site to layout https://t.me message widget, and iframe mysite. However JS cannot cross origin access iframe's scrollHeight due to https://t.me's server setup
3. (current preview). host my static site to layout telegram widgets and iframe this static site in DevHub. The static site looks good, for example: https://cg-msg-viewer.pages.dev/?group=NearZeroKnowledge. But when load it in DevHub, it doesn't layout right again, due to near social limitation. So I have to get messages in server side and find a frontend library to render it. Frontend telegram alone doesn't work because it requires user login to even list public group messages.
4. So I use telegram nodejs library based on mtproto, it can efficiently give a list of messages in [TL Schema Format](https://core.telegram.org/constructor/message), for example: https://j96g3uepe0.execute-api.us-east-1.amazonaws.com/messages/cryptonear. Then I'm trying to find a frontend library to render it. 
5. Official telegram web version Z and version K are both use non standard (self made frontend framework) that are very challenging to integrate with the frontend
6. Telegram-react, I'm able to run the Chat Message component, but the message format it takes from is in [tdlib format](https://core.telegram.org/tdlib/docs/classtd_1_1td__api_1_1message.html), which is not compatible to TL Schema Format above. And I found the conversion is non trivial.
7. Then I try to modify the service in 4. Although tdlib is more highlevel than mtproto, it's more difficult to use. tdlib's user session tokens is challenging to work with. Also tdlib's doesn't have official nodeJS binding, only browser JS, and non official node JS binding libraries have trouble in building and/or working on aws lambda. 

I plan to move to tdlib in another language to implement 7, then I will get the messages in right format for 6, and make a component and host on cloudflare pages, then iframe it here in DevHub.

**Comment by ailisp** [comment]
@frol Thanks! I'll create an issue on vm viewer.

> I believe that a page with content and two scrolls on it is better than no such page, so let's not go too deep into the custom rendering of Telegram messages.

Not sure if I read this correctly. Is this the resulting thing after integration integration iframe support?

> The current widget has reverse order of messages and uneven spacing

For ordering, it is intentional - want to display latest messages first to be consistent with posts. Do we want the last message last? @ori-near 

To clarify for others who come to discussion, the uneven spacing problem is the blocker we're trying to solve with all above comments.

**Comment by ailisp** [comment]
I found near social vm has iframe-resizer: https://github.com/NearSocial/VM/pull/29. However the support is incomplete and buggy, for example this issue: https://github.com/NearSocial/VM/issues/32. I've experimented several configurations of iframe-resizer on parent and on containing iframe, non of them could correctly obtain the message iframe height and pass to the parent :thinking:

**Comment by frol** [comment]
> > I believe that a page with content and two scrolls on it is better than no such page, so let's not go too deep into the custom rendering of Telegram messages.
>
> Not sure if I read this correctly. Is this the resulting thing after integration iframe support?

Yes, what I suggest is that we find a minimal working solution and run with it for now.

I also took a quick search for the ideal solution. I wish there is a standard widget for that, but Telegram Core only [offers widget for discussions](https://core.telegram.org/widgets/discussion) (a single message on a channel + comments) - this widget looks great, but it cannot preview a public group. [Web-clients](https://core.telegram.org/widgets/discussion) require authorization, and you cannot even preview a public group like https://web.telegram.org/k/#@neardev

**Comment by ailisp** [comment]
> [Web-clients](https://core.telegram.org/widgets/discussion) require authorization, and you cannot even preview a public group like https://web.telegram.org/k/#@neardev

What I am suggested in 6 & 7 above is essentially storing authorization on a aws lambda service and expose public group messages, then a modified web clients to render the messages. This is a working aws lambda in action: https://j96g3uepe0.execute-api.us-east-1.amazonaws.com/messages/cryptonear,
but the message format (mtproto) doesn't fit telegram-react (tdlib). It fits the official telegram web version K, but version K is using a self-made frontend framework and quite challenging to work with. I feel the other way, change the service to output into tdlib way and modify telegram-react seems to take less effort.

**Comment by frol** [comment]
@ailisp I wonder if [allowing more sandbox features](https://www.w3schools.com/tags/att_iframe_sandbox.asp) can fix the issue.

**Comment by ailisp** [comment]
@frol great suggestion! Experiment with a local near social viewer/VM, allow-same-origin is all I need to fix this issue. Will create a PR.

**Comment by ailisp** [comment]
I figure out an interesting way to render telegram messages well. Using an aws lambda function to crawl https://t.me/group_name/message-id, then take it as a div, combine divs to a single page dynamically, and iframe this dynamic page in near social. 

near social can correctly layout divs in an iframed page, but not iframe in iframe, as a result this way looks good:

image

**Comment by frol** [comment]
@ailisp I checked your preview link from the PR description and the background is black (is it intended?) and the font color in the messages is white (impossible to read messages since the background there is white):

image

https://near.social/#/bo.near/widget/gigs-board.pages.community.Telegram?label=zero-knowledge&tab=Telegram&nearDevGovGigsContractAccountId=devgovgigs.near

File: src/gigs-board/pages/community/Telegram.jsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+if (groupInfo === null || !groupInfo.ok) {
+  return "Loading ...";
+}
+
+const messageIds = groupInfo.body.messageIds;
+
+const Telegram = (
+  <div>
+    <iframe
+      iframeResizer={{ log: true }}
```
Comment (frol) [inline src/gigs-board/pages/community/Telegram.jsx:116]: Should we disable `log` now?

File: src/gigs-board/pages/community/Telegram.jsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      src={
+        "https://j96g3uepe0.execute-api.us-east-1.amazonaws.com/groups-ui/" +
+        group
+      }
+      frameborder="0"
+      style={{
+        width: "1px",
+        minWidth: "100%",
```
Comment (frol) [inline src/gigs-board/pages/community/Telegram.jsx:124]: Can you leave a comment about this quirk? Do we still need it, by the way?

File: src/gigs-board/pages/community/Telegram.jsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+const group = community.telegram;
+
+const groupInfo = fetch(
+  `https://j96g3uepe0.execute-api.us-east-1.amazonaws.com/groups/${group}`
+);
+if (groupInfo === null || !groupInfo.ok) {
+  return "Loading ...";
+}
+
+const messageIds = groupInfo.body.messageIds;
```
Comment (frol) [inline src/gigs-board/pages/community/Telegram.jsx:111]: It seems that `messageIds` and `groupInfo` is not used. Let's delete dead code

**Review by frol — APPROVED** [review]
There are some styling issues and minor comments to address, so I optimistically approve this PR, so you can merge this PR once all of the comments are addressed.

File: src/gigs-board/pages/community/Telegram.jsx (hunk 2 above)
Comment (ailisp) [reply src/gigs-board/pages/community/Telegram.jsx:124]: Yes, in this non-nested case, iframe resizer works, with these required by iframe-resizer

**Comment by ailisp** [comment]
> @ailisp I checked your preview link from the PR description and the background is black (is it intended?) and the font color in the messages is white (impossible to read messages since the background there is white):
> image
> 
> https://near.social/#/bo.near/widget/gigs-board.pages.community.Telegram?label=zero-knowledge&tab=Telegram&nearDevGovGigsContractAccountId=devgovgigs.near

Looks like you are in dark mode and the css is changing to black background in dark mode. I'll fixed it

**Pull request merged**